sdk/eventhub/Azure.Messaging.EventHubs/src/Processor/EventProcessorBase.cs

```diff
@@ -0,0 +1,729 @@
+                // we have no choice as we need to enforce balancing.  Otherwise, leave it as it is because the distribution wouldn't change.
+
+                if (!stealablePartitions.Any()
+                    && ownedPartitionsCount < minimumOwnedPartitionsCount)
+                {
+                    stealablePartitions = activeOwnership
+                        .Where(ownership => partitionDistribution[ownership.OwnerIdentifier] == maximumOwnedPartitionsCount)
+                        .Select(ownership => ownership.PartitionId);
+                }
+
+                // If any stealable partitions were found, randomly pick one of them to claim.
+
+                if (stealablePartitions.Any())
+                {
+                    var index = RandomNumberGenerator.Value.Next(stealablePartitions.Count());
+
+                    return ClaimOwnershipAsync(stealablePartitions.ElementAt(index), completeOwnershipEnumerable);
+                }
+            }
+
+            // No ownership has been claimed.
+
+            return Task.FromResult<PartitionOwnership>(null);
+        }
+
+        /// <summary>
+        ///   Stops an owned partition processing task in case it is running.  It is also removed from the tasks dictionary
+        ///   along with its corresponding token source.
+        /// </summary>
+        ///
+        /// <param name="partitionId">The identifier of the Event Hub partition whose processing is being stopped.</param>
+        /// <param name="reason">The reason why the processing for the specified partition is being stopped.</param>
+        ///
+        /// <returns>A task to be resolved on when the operation has completed.</returns>
+        ///
+        private async Task StopPartitionProcessingIfRunningAsync(string partitionId,
+                                                                 ProcessingStoppedReason reason)
+        {
+            // TODO: what if there is no token source?  It could happen if the user added a task by themself.
+
+            if (ActivePartitionProcessors.TryRemove(partitionId, out var processingTask)
+                && ActivePartitionProcessorTokenSources.TryRemove(partitionId, out var tokenSource))
+            {
+                try
+                {
+                    tokenSource.Cancel();
+                    await processingTask.ConfigureAwait(false);
+                }
+                catch (Exception)
+                {
+                    // TODO: delegate the exception handling to an Exception Callback.
+                }
+                finally
+                {
+                    tokenSource.Dispose();
+                }
+            }
+
+            // TODO: create the context only once instead of doing it every single time.
+            // TODO: if reason = Shutdown or OwnershipLost and we got an exception when closing, what should the final reason be?
+
+            var context = CreateContext(partitionId);
```

Comment:
I would expect the context to be very stable;  we should probably make the decision to consider it "create once and cache" or "create every time and users can cache if expensive to create" .... and document whichever approach we take.    I think either is justified, personally.

Reply:
I think we could just keep a context dictionary. That was my intent at first, but I wanted to get the draft out soon.